File: src/frontend/src/levels.ts
export type LevelDef = {
  level: number;
  scenarioId: number;
  title?: string;
};

export const LEVELS: LevelDef[] = [
  { level: 1, scenarioId: 1 },
  { level: 2, scenarioId: 1 },

  { level: 3, scenarioId: 2 },
  { level: 4, scenarioId: 2 },

  { level: 5, scenarioId: 3 },
  { level: 6, scenarioId: 3 },

  { level: 7, scenarioId: 4 },
  { level: 8, scenarioId: 4 },

  { level: 9, scenarioId: 5 },
  { level: 10, scenarioId: 5 },

  { level: 11, scenarioId: 6 },
  { level: 12, scenarioId: 7 },
  { level: 13, scenarioId: 8 },
  { level: 14, scenarioId: 9 },
  { level: 15, scenarioId: 6 },

  { level: 16, scenarioId: 6 },
  { level: 17, scenarioId: 7 },
  { level: 18, scenarioId: 8 },
  { level: 19, scenarioId: 9 },
  { level: 20, scenarioId: 6 },
];

File: src/backend/api/management/seeders/levels.py
from __future__ import annotations

from typing import Any

from django.db import transaction

from api.management.seeders.common import clamp_difficulty
from api.models import Email, Level, LevelEmail, Scenario


@transaction.atomic
def apply_curated_level_defs(defs: list[dict[str, Any]]) -> None:
    for ld in defs:
        lvl, _ = Level.objects.update_or_create(
            scenario=ld["scenario"],
            number=ld["number"],
            defaults=dict(title=ld["title"], briefing=ld["briefing"]),
        )

        LevelEmail.objects.filter(level=lvl).delete()

        base = ld.get("emails") or ld.get("base_emails") or []
        wave = ld.get("wave_emails") or []

        def upsert_email(em: dict[str, Any]) -> Email:
            email, _ = Email.objects.update_or_create(
                mode="simulation",
                scenario=ld["scenario"],
                sender_email=em["sender_email"],
                subject=em["subject"],
                defaults=dict(
                    sender_name=em["sender_name"],
                    body=em["body"],
                    is_phish=em["is_phish"],
                    difficulty=clamp_difficulty(em.get("difficulty", 1)),
                    category=em.get("category"),
                    links=em.get("links") or [],
                    attachments=em.get("attachments") or [],
                ),
            )
            email.full_clean()
            email.save()
            return email

        for idx, em in enumerate(base):
            email = upsert_email(em)
            LevelEmail.objects.create(level=lvl, email=email, sort_order=idx)

        for widx, em in enumerate(wave):
            email = upsert_email(em)
            LevelEmail.objects.create(level=lvl, email=email, sort_order=100 + widx)